lib/Parse/ParseDecl.cpp

```diff
@@ -6601,14 +6617,32 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
     }
   };
 
+  auto checkCodeCompletion = [&](SyntaxKind SK) -> bool {
+    if (Tok.is(tok::code_complete)) {
+      if (CodeCompletion)
+        CodeCompletion->completeInPrecedenceGroup(SK);
+      consumeToken();
+      return true;
+    }
+    return false;
+  };
+
+  // Skips the CC token if it comes without spacing.
+  auto skipPostfixCodeCompletionToken = [&](char lastTokTextEnd) {
+    if (Tok.is(tok::code_complete) && lastTokTextEnd != ' ' &&
+        lastTokTextEnd != '\t' && lastTokTextEnd != '\n')
```

Comment:
I don't think this works because `Tok.getText()` doesn't contain white spaces.
Usually, to determine any whitespaces precedes a token, `getEndOfPreviousLoc() != Tok.getLoc()` is sufficient.
Also, if we skip `code_complete` token, we should return `hasCodeCompletion()` result even if it's ignored because further parsing is meaningless.
So I suggest:
```cpp
auto skipPostfixCodeCompletionToken = [&] () -> Bool {
  if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc()) {
    conumeToken();
    return true;
  }
  return false;
}
...
if (skipPostfixCodeCompletionToken())
  return abortBody(/*hasCodeCompletion*/true);
```

Reply:
But is this really needed in the first place? Although 
```swift
precedencegroup MyPrecedence {
  associativity: right higherThan: AdditionPrecedence
}
```
is valid syntax, consecutive attributes in a single line is not a good style IMO.

Something like:
```cpp
while (Tok.isNot(tok::r_brace)) {
  if (Tok.is(tok::code_completion)) {
    // Suggest attribute names only if CC token is at the start of line.
    if (Tok.isAtStartOfLine() && CodeCompletion) {
      CodeCompletion->completePrecedenceGroup(...)
    }
    consumeToken(tok::code_complete);
    continue;
  }
  ...
```
without `skipPostfixCodeCompletionToken()` is sufficient.

Reply:
Placing the check before parsing the next attribute is a good move, though I think I've already tried and abandoned it for some reason. 

But, I believe that syntactic correctness is a superior criteria than 'bad style' for the compiler to support you.

> Usually, to determine any whitespaces precedes a token, `getEndOfPreviousLoc() != Tok.getLoc()` is sufficient. ... `if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc())`

Does `getEndOfPreviousLoc() != Tok.getLoc()` mean that there *is* or *isnt* a (`whitespace` | `\n` | `\r`)? Intuitively, the locations not matching tells me there is something in between, but not necessarily a symbol we are interested in.

Reply:
Between two adjacent tokens, only possible characters are whitespaces, newlines, comments, or other garbage characters *treated* as whitespaces (e.g. `U+00A0 No-Break Space`). Since comments are treated as a whitespace:
```swift
precedencegroup MyPrecedence {
  assignment: true/*comment*/higherThan: AssignmentPrecedence
}
```
is valid. So just checking (`whitespace` | `\n` | `\r`) is insufficient if we really want to check "whitespace"s.

Reply:
Right, I was actually just wondering if the second part of the condition from your reply had a typo (we should skip if there's, conversely, nothing in between: `getEndOfPreviousLoc() == Tok.getLoc()`)
```
if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc()) {
    conumeToken();
    return true;
}
```

Thanks for the tip on the preferred way of doing it.

Reply:
> should be `getEndOfPreviousLoc() == Tok.getLoc()`

Ah sorry. You are absolutely right!